multiple concurrency fixes

chappjc opened the pull request:
The primary issue addressed in this PR is the lack of concurrent access
control to websocket's read methods.  Locking for the write methods was
recently added, but read methods have the same requirement.  From
https://godoc.org/github.com/gorilla/websocket#hdr-Concurrency:

>  Connections support one concurrent reader and one concurrent writer.
>
> Applications are responsible for ensuring that no more than one goroutine calls the write methods (NextWriter, SetWriteDeadline, WriteMessage, WriteJSON, EnableWriteCompression, SetCompressionLevel) concurrently **and that no more than one goroutine calls the read methods** (NextReader, SetReadDeadline, ReadMessage, ReadJSON, SetPongHandler, SetPingHandler) concurrently.
>
> The Close and WriteControl methods can be called concurrently with all other methods.

Similar to how `wrapper.writeLocker` is handled (unlocking on close), this adds a `readLocker` with identical semantics except for reads.  Note that readers from `wrapper.NextReader` must be `Close`d now.

This PR also eliminates the use in `session` of the `reader` and `writer` defined in util.go since it is redundant with the locking in `transport/websocket.wrapper`.

Details:

**session**: `upgradeLocker` is for protecting the `session.conn` field,
not concurrent writes on the actual connection. Only use it that way.

Remove `writeLocker` from `session` since they are in the `wrapper`.

No longer use `util.go`'s `newWriter` and `newReader` from util.go since
the locking is in the `wrapper`.

**util.go**: Changes even though util.go is no longer used.

Use a full write lock, not read lock, with `reader.Close` and `reader.Read`.
With RLock and RUnlock, concurrent reads were not prevented, but this
is a requirement of gorilla's connection read methods too.

Change the locker type in both reader and writer from `*sync.RWMutex`
to `sync.Locker`, which is satisfied by both `Mutex` and `RWMutex`.

Note: util.go is now UNUSED.

**transport/websocket**: `wrapper` needs locking around reads too.

Add `wrapper.readLocker`, and use it in `NextReader`.
Add `rcWrapper`, similar to `wcWrapper`, for unlocking the `wrapper`'s read
lock on `Close`.

**packet**: A failed read from a decoder must close the reader.

As in `(*encoder).NextWriter`, which calls the writer's `Close` method when
a `Write` fails, `(*decoder).NextReader` must call the reader's `Close` method
on a failed read (`io.ReadFull`).

**bonus**: go.mod needs a go version line since Go 1.13 adds one

chappjc commented on packet/decoder.go:
```diff
@@ -23,6 +23,7 @@ func (e *decoder) NextReader() (base.FrameType, base.PacketType, io.ReadCloser,
 	}
 	var b [1]byte
 	if _, err := io.ReadFull(r, b[:]); err != nil {
+		r.Close()
```
unlocks the FrameReader

chappjc commented on session.go:
```diff
@@ -19,24 +19,28 @@ type session struct {
 	closeOnce sync.Once
 	context   interface{}
 
-	writeLocker   sync.RWMutex
```
Connection r/w locking is in `wrapper`.

chappjc commented on session.go:
```diff
@@ -19,24 +19,28 @@ type session struct {
+		return nil, err
+	}
+
+	m.Add(ses)
```
Only add the session if setdeadline succeeds.

chappjc commented on session.go:
```diff
@@ -66,39 +70,56 @@ func (s *session) Close() error {
-				io.Copy(w, r)
-				return w.Close()
+				// echo?
+				_, err = io.Copy(w, r)
```
This is the more important error to return.

chappjc commented on session.go:
```diff
@@ -128,61 +149,50 @@ func (s *session) RemoteHeader() http.Header {
 			return 0, 0, nil, err
 		}
-		s.upgradeLocker.RUnlock()
-		return ft, pt, newReader(r, &s.upgradeLocker), nil
```
No need for the reader in util.go.

chappjc commented on transport/websocket/wrapper.go:
```diff
@@ -13,30 +13,46 @@ import (
 type wrapper struct {
 	*websocket.Conn
 	writeLocker *sync.Mutex
+	readLocker  *sync.Mutex
```
concurrent reads are not OK either.

chappjc commented on transport/websocket/wrapper.go:
```diff
@@ -13,30 +13,46 @@ import (
+}
+
+func (r rcWrapper) Close() error {
+	_, err := io.Copy(ioutil.Discard, r)
```
a la util.go's Close

chappjc commented on session.go:
```diff
@@ -128,61 +150,50 @@ func (s *session) RemoteHeader() http.Header {
+	defer s.upgradeLocker.RUnlock()
 
-		if same {
-			return
```
This look screamed "let's race", but perhaps I don't understand the intention here.

chappjc commented on session.go:
```diff
@@ -128,61 +150,50 @@ func (s *session) RemoteHeader() http.Header {
 			return 0, 0, nil, err
 		}
-		s.upgradeLocker.RUnlock()
-		return ft, pt, newReader(r, &s.upgradeLocker), nil
```
Note: Using `upgradeLocker` for the reader's lock was an improper (re)use of the `upgradeLocker` mutex which appears to be intended to only protect the `conn` and `transport` fields of the `session` struct, rather than also protect concurrent writes on the websocket connection.

But this is irrelevant since the `reader`/`newReader` stuff in util.go is unnecessary given `transport/websocket.wrapper`.

erkie commented:
This looks amazing @chappjc thanks for the hard work. My policy for this project is to really just merge most things that people seem to have put a lot of effort into. I'm happy to merge this ASAP. There was a TODO there, do you want to fix that? I'd recommend leaving it and opening a separate PR for that issue, just to get things moving.

chappjc commented:
I'll leave the TODO for now.

One thing that I don't love about the Close-to-unlock approach that this PR expands on is that it might lead to deadlocks in code that is presently neglecting to Close.  There were several such cases in tests and demos. I just fixed some more in commit https://github.com/googollee/go-engine.io/pull/78/commits/cc7734cb937fc331c1fffdb38b6fa2f23bbbe315 this morning.

To help with that, I added a nag message in https://github.com/googollee/go-engine.io/pull/78/commits/0167b3d088682e186f15fe4bbd920ff3964a7426 to print a message to stdout if the ReadCloser or WriteCloser weren't Closed within a fair amount of time.  I don't love this either, but it might head off bug reports due to a forgotten Close.

The last thing I'm worried about is two different intermittent failures in `TestEngineUpgrade`.  Ignoring the legit race on the actual `*testing.T` object when two goroutines try to to `t.FailNow`, one of the intermittent failures is:

> === RUN   TestEngineUpgrade
2019/09/22 11:08:50 httptest.Server blocked in Close after 5 seconds, waiting for connections:
  *net.TCPConn 0xc00016e000 127.0.0.1:41350 in state active

I don't know why it only happens occasionally, but it suggests that a connection can sometimes stay  open in this test.  Note that this happens on `master` too, so it probably not an issue with the present changes.

The other intermittent failure is an unexpected `EOF`, but I don't get what could cause it:

```
--- FAIL: TestEngineUpgrade (0.00s)
    require.go:765: 
        	Error Trace:	engine_test.go:197
        	            				asm_amd64.s:1357
        	Error:      	Expected nil, but got: &errors.errorString{s:"EOF"}
        	Test:       	TestEngineUpgrade
    engine_test.go:282: 
        	Error Trace:	engine_test.go:282
        	Error:      	Expected nil, but got: &websocket.CloseError{Code:1006, Text:"unexpected EOF"}
        	Test:       	TestEngineUpgrade
```

That one I do not get on `master`, but it feels like a manifestation of an existing issue.  Not sure.

Using `-count 1` to force rerunning a test without code changes:

    go test -timeout 40s -run '^(TestEngineUpgrade)$' -v -count 1 .

erkie commented:
@chappjc nice work and summary of the state of things. Would you feel cool with this getting merged so it can be tried out by a broader set of users? There seems to be a few intermittent issues in master, hopefully this PR will make it easier to find those. If you're ok with that I'll merge.

chappjc commented:
Yeah, that sounds good.  Thanks.
